pkg/spade-spike/spade-spike.go

```diff
@@ -0,0 +1,289 @@
+			logger.Info("Block not found")
+			close(notFound)
+		}
+	}})
+	defer bswap.Close()
+	defer network.Stop()
+	connectContext, cancel := context.WithTimeout(parent, c.timeout)
+	defer cancel()
+	logger.Info("Connecting to target peer...")
+	err := c.host.Connect(connectContext, target)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to connect to target peer")
+	}
+
+	resultChan := make(chan blocks.Block)
+	errChan := make(chan error)
+	go func() {
+		logger.Debug("Retrieving block...")
+		blk, err := bswap.GetBlock(connectContext, cid)
+		if err != nil {
+			logger.Info(err)
+			errChan <- err
+		} else {
+			resultChan <- blk
+		}
+	}()
+	select {
+	case <-notFound:
+		return nil, errors.New("DONT_HAVE received from the target peer")
+
+	case blk := <-resultChan:
+		return blk, nil
+
+	case err := <-errChan:
+		return nil, errors.Wrap(err, "error received %s")
+	}
+}
+
+// Starts with the root CID, then fetches a random CID from the children and grandchildren nodes, until it reaches `traverseDepth`
+// Note: the root CID is considered depth `0`, so passing `traverseDepth=0` will only fetch the root CID
+// Returns true if all retrievals were successful, false if any failed
+func SpadeTraversal(ctx context.Context, startingCid goCid.Cid, p peer.AddrInfo, traverseDepth uint) (bool, error) {
+	cidToRetrieve := startingCid
+
+	// support structures such as: https://github.com/filecoin-project/go-dagaggregator-unixfs#grouping-unixfs-structure
+	for i := uint(0); i <= traverseDepth; i++ {
+		// For some reason, need to re-init the host & client every time we do a fetch
+		// otherwise, we get context timeout error after the first fetch
+		host, err := net.InitHost(ctx, nil)
+		if err != nil {
+			return false, errors.Wrap(err, "failed to init host %s")
+		}
+
+		client := NewBitswapClient(host, time.Second*1)
+
+		// Retrieval
+		logger.Debugf("retrieving %s\n", cidToRetrieve.String())
+		blk, err := client.RetrieveBlock(ctx, p, cidToRetrieve)
+		if err != nil {
+			return false, errors.Wrap(err, "unable to retrieve cid %s")
+		}
+
+		if i == traverseDepth {
+			// we've reached the bottom of the tree
+			logger.Debugf("retrieved data cid %s which contains %d bytes\n", cidToRetrieve.String(), len(blk.RawData()))
+			return true, nil
+		}
+
+		// if not at bottom of the tree, keep going down the links until we reach it
+		links, err := FindLinks(ctx, blk)
+		if err != nil {
+			return false, errors.Wrap(err, "unable to find links")
+		}
+
+		logger.Debugf("retrieved %s which has %d links\n", cidToRetrieve.String(), len(links))
+
+		nextIndex := 0
+		rand.Seed(time.Now().UnixNano())
+		if i == 0 {
+			if len(links) < 2 {
+				return false, errors.New("starting node contains less than 2 links, will not traverse any further")
+			}
+
+			// from the starting node's children, never grab the first link as it refers to the AggregateManifest
+			nextIndex = 1 + rand.Intn(len(links)-1)
```

Comment:
@ribasushi re: not hardcoding dag-aggregator stuff - it seems like this is also an issue too, since we have to hard-code that the _first child of the root_ is reserved for the AggregateManifest. Do you think there's a better way of doing this that's more generalized? I.e, do we have another way of knowing something is a `manifest` file and shouldn'
t be fetched?

Reply:
🤷 do two challenges per level. I.e. accept a parameter for that as well. Be creative ;)